sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -516,29 +542,62 @@ pi_result piextGetDeviceFunctionPointer(pi_device device, pi_program program,
     return cast<pi_result>(ret_err);
   }
 
-  using FuncT =
-      cl_int(CL_API_CALL *)(cl_device_id, cl_program, const char *, cl_ulong *);
-
-  // TODO: add check that device supports corresponding extension
-  FuncT func_ptr =
-      reinterpret_cast<FuncT>(clGetExtensionFunctionAddressForPlatform(
-          cast<cl_platform_id>(platform), "clGetDeviceFunctionPointerINTEL"));
-  // TODO: once we have check that device supports corresponding extension,
-  // we can insert an assertion that func_ptr is not nullptr. For now, let's
-  // just return an error if failed to query such function
-  // assert(
-  //     func_ptr != nullptr &&
-  //     "Failed to get address of clGetDeviceFunctionPointerINTEL function");
-
-  if (!func_ptr) {
-    if (function_pointer_ret)
-      *function_pointer_ret = 0;
-    return PI_INVALID_DEVICE;
+  // Get device supported extensions
+  size_t extension_size;
+  clGetDeviceInfo(cast<cl_device_id>(device), CL_DEVICE_EXTENSIONS, 0, nullptr,
+                  &extension_size);
+  char *extensions = new char[extension_size];
+  clGetDeviceInfo(cast<cl_device_id>(device), CL_DEVICE_EXTENSIONS,
+                  extension_size, extensions, nullptr);
+
+  // Check if clGetDeviceFunctionPointer is in list of extensions
+  if (is_in_separated_string(extensions, ' ',
+                             "clGetDeviceFunctionPointerINTEL")) {
```

Comment:
According to the OpenCL specification [`clGetExtensionFunctionAddressForPlatform`](https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clGetExtensionFunctionAddressForPlatform.html) should return null if the function cannot be found or if the platform is invalid. Could this be used as a way to identify whether or not the extension is supported?

As @smaslov-intel suggested you should use `getExtFuncFromContext` instead, which will return `PI_INVALID_VALUE` if `clGetExtensionFunctionAddressForPlatform` returned nullptr.

@smaslov-intel - Currently `getExtFuncFromContext` only caches found functions. Should we consider caching nullptr as well to remember and report missing extension functions?

Reply:
I have tested this approach just now. I am having clGetExtensionFunctionAddressForPlatform in getExtFuncFromContext return 1 instead of nullptr. So it returns a false positive in this case.

Reply:
`getExtFuncFromContext` should return `PI_INVALID_VALUE` if it fails to find the extension function, in which case you don't have to care about the returned pointer.

Reply:
It will only return `PI_INVALID_VALUE` if the function is nullptr
```
  T FuncPtr =
      (T)clGetExtensionFunctionAddressForPlatform(curPlatform, FuncName);

  if (!FuncPtr)
    return PI_INVALID_VALUE;
```

Reply:
> I am having clGetExtensionFunctionAddressForPlatform in getExtFuncFromContext return 1 instead of nullptr. So it returns a false positive in this case.

Is this a bug in OpenCL implementation? We need to confirm with the owners.

> Currently getExtFuncFromContext only caches found functions. Should we consider caching nullptr as well to remember and report missing extension functions?

Yes, we should add caching of extensions known to be not present. It is an overkill to query it again and again in piextGetDeviceFunctionPointer.

Reply:
@AidanBeltonS : are you going to change the code to use `getExtFuncFromContext` ?

Reply:
I have the function using just getExtFuncFromContext now. It also caches nullptr functions. 

I have figured out the error I was getting, the extensions which end with INTEL do not show up when you query a device or platform for its extensions. Fallback search does need to be used before using the function as the opencl cpu runtime throws an exception with unknown names. GPU's dont have this problem but do sometimes not see the function, similar to level zero. 

Let me know if this behaviour is that you are looking for. Thanks